### PR TITLE
fix: `TypeAnnotation` with comments generates incorrect code

### LINKED ISSUE #14751: [Bug]: Babel produces invalid TypeScript code if comment follows type annotation and is not stripped
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```ts
export default (): void /* hi! */ => {};
```

### Configuration file name

babel.config.json

### Configuration

```json
{}
```

### Current and expected behavior

Using this test harness:
```ts
const babel = require('@babel/core');
const result = babel.transformSync("export default (): void /* hi! */ => {};", {
    filename: './source.ts',
    parserOpts: {
        plugins: ["typescript"],
    },
});
console.log(result.code);
```

The resulting code produced by Babel looks like this:

```ts
export default ((): void
/* hi! */
=> {});
```

This is invalid TypeScript/JavaScript as the arrow function is not allowed to have a line terminator preceding it.

Removing the comment from the source produces working output:

```ts
export default ((): void => {});
```

### Environment

  ```
System:
    OS: macOS 11.6.7
  Binaries:
    Node: 16.15.1
    Yarn: 1.22.10
    npm: 8.11.0
  npmPackages:
    @babel/core: ^7.18.6 => 7.18.6 
```

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14751 
| Patch: Bug Fix?          |√
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR also contains some very minor performance optimizations.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14762"></a>

### CODE REVIEW

File: packages/babel-generator/src/generators/methods.ts [hunk 1]
```diff
@@ -11,15 +11,7 @@ export function _params(
...
-    if (node.type === "ArrowFunctionExpression") {
-      this._noLineTerminator = true;
-      this.print(node.returnType, node);
-      this._noLineTerminator = false;
-    } else {
-      this.print(node.returnType, node);
-    }
-  }
+  this.print(node.returnType, node);
 }
```
Comment (liuxingbaoyu) [inline packages/babel-generator/src/generators/methods.ts:15]: Reverted here https://github.com/babel/babel/pull/14758.

**Comment by liuxingbaoyu** [comment]
There seem to be two ways, one is like https://github.com/babel/babel/pull/14758 and the other is using `_printStack` in the `print` method, what do you think?
In addition this problem seems to exist in the flow, do we still need to consider it? (I don't know anything about flow, it will need your detailed guidance)

**Comment by nicolo-ribaudo** [comment]
I prefer the https://github.com/babel/babel/pull/14758 approach, since it lets us get a "better" output where we can print newlines.

The flow rules are the same, except that the AST nodes are `ArrayTypeAnnotation`, `IndexedAccessType` and `GenericTypeAnnotation` (you can test the example above on astexplorer.org).

File: packages/babel-generator/test/fixtures/typescript/types-with-comments-retainLines/output.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+/* 1 */
+C /* 2 */[
+/* 3 */
+0];
+
+type T2 = U.
+/* 1 */
+C /* 2 */[];
+
+
```
Comment (liuxingbaoyu) [inline packages/babel-generator/test/fixtures/typescript/types-with-comments-retainLines/output.js:11]: There seems to be a bug here.

Comment (nicolo-ribaudo) [reply packages/babel-generator/test/fixtures/typescript/types-with-comments-retainLines/output.js:11]: Yes, the `/* 3 */` comment should be printed _somewhere_. It's a bug that we already have regardless of this PR.

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Review by JLHwung — APPROVED** [review]

**Pull request merged**